rust/lance/src/io/writer/statistics.rs

```diff
@@ -150,12 +437,300 @@ fn get_boolean_statistics(arrays: &[&ArrayRef]) -> StatisticsRow {
+                max_length = std::cmp::max(max_length, row.len());
+                min_length = std::cmp::min(min_length, row.len());
+                let values = as_primitive_array::<T>(&row);
+                for (i, val) in values.iter().enumerate() {
+                    if let Some(val) = val {
+                        if let Some(Ordering::Greater) = val.partial_cmp(&max_value[i]) {
+                            max_value[i] = val;
+                        }
+                        if let Some(Ordering::Less) = val.partial_cmp(&min_value[i]) {
+                            min_value[i] = val;
+                        }
+                    }
+                    if i == max_length {
+                        break;
+                    }
+                }
+            } else {
+                continue;
+            }
+        }
+    });
+
+    let DataType::List(f) = arrays[0].data_type() else {
+        todo!()
+    };
+    let min_value = min_value[0..min_length]
+        .iter()
+        .map(|&x| ScalarValue::from(x))
+        .collect::<Vec<_>>();
+    let max_value = max_value[0..max_length]
+        .iter()
+        .map(|&x| ScalarValue::from(x))
+        .collect::<Vec<_>>();
+    StatisticsRow {
+        null_count: ScalarValue::Int64(Some(null_count)),
+        min_value: ScalarValue::List(Some(min_value), f.clone()),
+        max_value: ScalarValue::List(Some(max_value), f.clone()),
+    }
+}
+
+fn cast_dictionary_arrays<'a, T: ArrowDictionaryKeyType + 'static>(
+    arrays: &'a [&'a ArrayRef],
+) -> Vec<&Arc<dyn Array>> {
+    arrays
+        .iter()
+        .map(|x| x.as_dictionary::<T>().values())
+        .collect::<Vec<_>>()
+}
+
+fn get_dictionary_statistics(arrays: &[&ArrayRef]) -> StatisticsRow {
+    let data_type = arrays[0].data_type();
+    match data_type {
+        DataType::Dictionary(key_type, _) => match key_type.as_ref() {
+            DataType::Int8 => collect_statistics(&cast_dictionary_arrays::<Int8Type>(arrays)),
+            DataType::Int16 => collect_statistics(&cast_dictionary_arrays::<Int16Type>(arrays)),
+            DataType::Int32 => collect_statistics(&cast_dictionary_arrays::<Int32Type>(arrays)),
+            DataType::Int64 => collect_statistics(&cast_dictionary_arrays::<Int64Type>(arrays)),
+            DataType::UInt8 => collect_statistics(&cast_dictionary_arrays::<UInt8Type>(arrays)),
+            DataType::UInt16 => collect_statistics(&cast_dictionary_arrays::<UInt16Type>(arrays)),
+            DataType::UInt32 => collect_statistics(&cast_dictionary_arrays::<UInt32Type>(arrays)),
+            DataType::UInt64 => collect_statistics(&cast_dictionary_arrays::<UInt64Type>(arrays)),
+            _ => {
+                panic!("Unsupported dictionary key type: {}", key_type);
+            }
+        },
+        _ => {
+            panic!("Unsupported data type for dictionary: {}", data_type);
+        }
+    }
+}
+
+fn get_struct_statistics(arrays: &[&ArrayRef]) -> StatisticsRow {
```

Comment:
We don't need to construct statistics for structs. We will visit all leaf fields in depth-first order to get all the child fields.

Reply:
Removed.